unit_tests/stream_iteration.cpp

```diff
@@ -0,0 +1,293 @@
+      this->invocations_cons++; 
+      this->out += val;
+    }
+  );
+  this->check_no_composed();
+}
+/*
+TYPED_TEST(stream_iteration_test, poly_no_composed)
+{ 
+  this->setup_no_composed();
+  grppi::stream_iteration(this->poly_execution_,
+    [this]() -> optional<int> {
+      this->invocations_gen++;
+      if (this->count < this->n){
+       this->count+=1;
+       return 1;
+      }
+      else return {};
+    },
+    [this](int val){
+      this->invocations_oper++;
+      return val+1;
+    },
+    [this](int val) {
+      this->invocations_pred++;
+      return val<10;
+    },
+    [this](int val) {
+      this->invocations_cons++;
+      this->out += val;
+    }
+  );
+  this->check_no_composed();
+}
+*/
+TYPED_TEST(stream_iteration_test, static_composed_pipeline)
+{
+  this->setup_composed_pipeline();
+  grppi::stream_iteration(this->execution_,
+    [this]() -> optional<int> {
+      this->invocations_gen++;
+      if (this->count < this->n) {
+       this->count+=1;
+       return 1;
+      }
+      else return {};
+    },
+    grppi::pipeline(this->execution_,
+       [this](int val){
+         this->invocations_stage1++;
+         return val+1;
+       },
+       [this](int val){
+         this->invocations_stage2++;
+         return val+1;
+       }
+    ),
+    [this](int val) {
+      this->invocations_pred++;
+      return val<10;
+    },
+    [this](int val) {
+      this->invocations_cons++;
+      this->out += val;
+    }
+  );
+  this->check_composed_pipeline();
+}
+/*
+TYPED_TEST(stream_iteration_test, poly_composed_pipeline)
+{
+  this->setup_composed_pipeline();
+  grppi::stream_iteration(this->poly_execution_,
+    [this]() -> optional<int> {
+      this->invocations_gen++;
+      if (this->count < this->n) {
+       this->count+=1;
+       return 1;
+      }
+      else return {};
+    },
+    grppi::pipeline(this->execution_,
+       [this](int val){
+         this->invocations_stage1++;
+         return val+1;
+       },
+       [this](int val){
+         this->invocations_stage2++;
+         return val+1;
+       }
+    ),
+    [this](int val) {
+      this->invocations_pred++;
+      return val<10;
+    },
+    [this](int val) {
+      this->invocations_cons++;
+      this->out += val;
+    }
+  );
+  this->check_composed_pipeline();
+
+}
+*/
+TYPED_TEST(stream_iteration_test, static_composed_farm)
+{
+  this->setup_composed_farm();
+  grppi::stream_iteration(this->execution_,
+    [this]() -> optional<int> {
+      this->invocations_gen++;
+      if (this->count < this->n) {
+       this->count+=1;
+       return 1;
+      }
+      else return {};
+    },
+    grppi::farm(this->execution_,
+       [this](int val){
+         this->invocations_oper++;
+         return val+1;
+       }
+    ),
+    [this](int val) {
+      this->invocations_pred++;
+      return val<10;
+    },
+    [this](int val) {
+      this->invocations_cons++;
+      this->out += val;
+    }
+  );
+  this->check_composed_farm();
+}
+/*
```

Comment:
This is commented because de polymorphic has not the stream iteration implemented yet.